File: SLP em java/src/SLP.java
import java.util.ArrayList;
import java.util.Random;

public class SLP {

	private static SLP instance;
	private int limiar;
	private double taxaDeAprendizado;
	private int numIteracoesTreino;
	private ArrayList<Float> pesos;
	
	private SLP() {
		this.limiar = 3500;
		this.taxaDeAprendizado = 0.1;
		this.numIteracoesTreino = 100;
		this.pesos = new ArrayList<Float>();
	}
	
	public static SLP getInstance() {
		if (instance == null)
			instance = new SLP();
		return instance;
	}
	
	private void inicializaPesos() {
		Random r = new Random();
		for (int i = 0; i < 256; i++)
			pesos.add((float) (r.nextInt(9) + 1));
	}
	
	private int avaliarEntrada(int [] histograma) {
		float valor = 0;
		for (int i = 0; i < 256; i++)
			valor += (pesos.get(i) * histograma[i]);
		return (valor >= limiar)? 1 : 0;
	}
	
	private float calcularNovoPeso(float peso, int saidaEsperada, int saidaReal, int xi) {
		float novoPeso = (float) (peso + (taxaDeAprendizado * (saidaEsperada - saidaReal) * xi));
		return novoPeso;
	}
	
	private void treinarPerceptronComEntrada(int [] histograma) {
		int saidaEsperada = histograma[256];
		int saidaReal = avaliarEntrada(histograma);
		if (saidaReal != saidaEsperada) {
			for (int i = 0; i < 256; i++)
				pesos.set(i, calcularNovoPeso(pesos.get(i), saidaEsperada, saidaReal, histograma[i]));
		}
	}
	
	private void treinarPerceptronComDataset(int [][] dataset) {
		for (int i = 0; i < dataset.length; i++)
			treinarPerceptronComEntrada(dataset[i]);
	}
	
	private int [] testarComDataset(int [][] dataset) {
		int [] resultados = new int [dataset.length];
		for (int i = 0; i < dataset.length; i++)
			resultados[i] = avaliarEntrada(dataset[i]);
		return resultados;
	}
	
	private void iteracoesDeTreino(int [][] dataset) {
		int [] resultadosAntes = new int [dataset.length];
		int [] resultadosDepois = new int [dataset.length];
		boolean iguais = true;
		for (int i = 0; i < numIteracoesTreino; i++) {
			resultadosAntes = testarComDataset(dataset);
			treinarPerceptronComDataset(dataset);
			resultadosDepois = testarComDataset(dataset);
			for (int j = 0; j < dataset.length && iguais; j++) {
				if (resultadosAntes[j] != resultadosDepois[j]) {
					iguais = false;
				}
			}
			if (iguais) {
				System.out.println("Pesos convergiram, iterações necessárias: " + i + ".");
				break;
			}
		}
	}
	
	private String quantificarEntradas(int [][] dataset) {
		int cavalos = 0;
		int naoCavalos = 0;
		for (int i = 0; i < dataset.length; i++) {
			if (dataset[i][256] == 1)
				cavalos++;
			else
				naoCavalos++;
		}
		return "\n\tCavalo: " + cavalos + "\n\tNão Cavalos: " + naoCavalos + "\n";
	}
	
	private float percentualDeAcerto(int [][] datasetTestes, int [] resultados) {
		int acertos = 0;
		for (int i = 0; i < datasetTestes.length; i++) {
			if (datasetTestes[i][256] == resultados[i])
				acertos++;
		}
		return (float) acertos/datasetTestes.length;
	}
	
	public void slp(int [][] datasetDeTreino, int[][] datasetDeTestes) {
		inicializaPesos();
		
		System.out.println("Imagens para treino: " + datasetDeTreino.length);
		System.out.println("Imagens para teste: " + datasetDeTestes.length);

		System.out.println("Quantificar entradas treino: " + quantificarEntradas(datasetDeTreino));
		System.out.println("Quantificar entradas testes: " + quantificarEntradas(datasetDeTestes));
		
		System.out.print("Pesos antes do treino: ");
		for (int i = 0; i < pesos.size() - 1; i++)
			System.out.print(pesos.get(i) + ", ");
		System.out.println(pesos.get(pesos.size() - 1) + ".");
		
		iteracoesDeTreino(datasetDeTreino);
		
		System.out.print("Pesos após o treino: ");
		for (int i = 0; i < pesos.size() - 1; i++)
			System.out.print(pesos.get(i) + ", ");
		System.out.println(pesos.get(pesos.size() - 1) + ".");

		int [] resultados = testarComDataset(datasetDeTestes);
		System.out.print("Resultados: ");
		for (int i = 0; i < resultados.length - 1; i++)
			System.out.print(resultados[i] + ", ");
		System.out.println(resultados[resultados.length - 1] + ".");
		
		System.out.println("Percentual de acerto: " + percentualDeAcerto(datasetDeTestes, resultados) * 100 + "%");
	}
	
}

File: SLP.py
from random import randint
from datasets import *


def executar_programa():

    def iniciar_pesos():
        lista_de_pesos = []
        for x in range(256):
            lista_de_pesos.append(randint(1, 10))  # Inicia os pesos de cada entrada aleatoriamente com valores entre 1 e 10
        return lista_de_pesos

    def avaliar_entrada(histograma, pesos, limiar):  # Funcao ativacao
        valor = 0
        for y in range(256):
            valor = valor + pesos[y] * histograma[y]

        if valor >= limiar:
            return 1
        else:
            return 0

    def iteracoes_de_treino(dataset, pesos, taxa_de_aprendizado, limiar, iteracoes): # Repete o treino por n iteracoes, ou ate os pesos convergirem

        def calcular_novo_peso(peso, saida_esperada, saida_real, taxa_de_aprendizado, xi):  # Atualiza o um peso caso haja erro durante o treino
            novo_peso = peso + (taxa_de_aprendizado * (saida_esperada - saida_real) * xi)
            return novo_peso

        def treinar_perceptron_com_entrada(histograma, pesos, taxa_de_aprendizado, limiar):  # Executa o treino para uma entrada
            saida_esperada = histograma[256]
            saida_real = avaliar_entrada(histograma, pesos, limiar)
            if saida_real != saida_esperada:  # Recalcula o peso se houver erro
                for x in range(len(pesos)):
                    pesos[x] = calcular_novo_peso(pesos[x], saida_esperada, saida_real, taxa_de_aprendizado, histograma[x])
            return pesos

        def treinar_perceptron_com_dataset(dataset, pesos, taxa_de_aprendizado, limiar):  # treina o perceptron com um dataset inteiro
            for x in range(len(dataset)):
                pesos = treinar_perceptron_com_entrada(dataset[x], pesos, taxa_de_aprendizado, limiar)
            return pesos

        x = 0
        while x < iteracoes:
            resutados_antes = testar_com_dataset(dataset, pesos, limiar)
            pesos = treinar_perceptron_com_dataset(dataset, pesos, taxa_de_aprendizado, limiar)
            resultados_depois = testar_com_dataset(dataset, pesos, limiar)
            if resutados_antes == resultados_depois:
                print("Pesos convergiram, iteracoes necessarias: " + str(x))
                break
            x = x + 1
        return pesos

    def testar_com_dataset(dataset, pesos, limiar): # gera as classificacoes de um dataset
        resultados = []
        for x in range(len(dataset)):
            resultados.append(avaliar_entrada(dataset[x], pesos, limiar))
        return resultados

    def percentual_de_acerto(dataset_testes, resultados): # Calcula o percentual de acertos de um dataset com os resultados do teste
        acertos = 0
        for x in range(len(dataset_testes)):
            if dataset_testes[x][256] == resultados[x]:
                acertos = acertos + 1
        return acertos/len(dataset_testes)

    def quantificar_entradas(dataset):
        cavalos = 0
        nao_cavalos = 0
        for x in range(len(dataset)):
            if dataset[x][256] == 1:
                cavalos = cavalos + 1
            else:
                nao_cavalos = nao_cavalos + 1
        return "Cavalos: " + str(cavalos) + " | Nao Cavalos: " + str(nao_cavalos)


    dataset_de_treino, dataset_de_testes = getDatasets()

    # dataset[n][256] == classe da imagem n, 0 == Nao_Cavalo, 1 == Cavalo

    limiar = 3500  # Valor escolhido para ser comparado na funcao de aticavacao
    taxa_de_aprendizado = 0.1
    num_iteracoes_treino = 100
    pesos = iniciar_pesos()
    print("Imagens para treino: " + str(len(dataset_de_treino)))
    print("Imagens para testes: " + str(len(dataset_de_testes)))
    print("Treino(" + quantificar_entradas(dataset_de_treino) + ")")
    print("Testes(" + quantificar_entradas(dataset_de_testes) + ")")
    print("Pesos antes do treino: " + str(pesos) + "\nTreinando...")
    print("Pesos apos o treino: " + str(iteracoes_de_treino(dataset_de_treino, pesos, taxa_de_aprendizado, limiar, num_iteracoes_treino)))
    print("Resultados: " + str(testar_com_dataset(dataset_de_testes, pesos, limiar)))
    print("% de acerto: " + str(percentual_de_acerto(dataset_de_testes, testar_com_dataset(dataset_de_testes, pesos, limiar))))


executar_programa()
